src/modules/audio/openal/DeviceWatcher.cpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) 2006-2022 LOVE Development Team
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software. If you use this software
+ *    in a product, an acknowledgment in the product documentation would be
+ *    appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ **/
+
+#include <chrono>
+
+// LOVE
+#include "common/delay.h"
+#include "DeviceWatcher.h"
+
+#ifndef ALC_CONNECTED
+#define ALC_CONNECTED 0x313
+#endif
+
+#ifndef ALC_ALL_DEVICES_SPECIFIER
+#define ALC_ALL_DEVICES_SPECIFIER 0x1013
+#endif
+
+namespace love
+{
+namespace audio
+{
+namespace openal
+{
+
+DeviceWatcher::DeviceWatcher(ALCdevice *device, const std::vector<ALint> &attribs)
+: deviceChanged(false)
+, stopped(false)
+, device(device)
+, attribs(attribs)
+, alcReopenDeviceSOFT(nullptr)
+{
+	if (
+		alcIsExtensionPresent(nullptr, "ALC_ENUMERATE_ALL_EXT") &&
+		alcIsExtensionPresent(device, "ALC_EXT_disconnect") &&
+		alcIsExtensionPresent(device, "ALC_SOFT_reopen_device"))
+	{
+		alcReopenDeviceSOFT = (LPALCREOPENDEVICESOFT) alcGetProcAddress(nullptr, "alcReopenDeviceSOFT");
+		currentDevice = getFirstDevice();
+	}
+}
+
+DeviceWatcher::~DeviceWatcher()
+{
+}
+
+void DeviceWatcher::stop()
+{
+	stopped = true;
+}
+
+bool DeviceWatcher::tryReconnect(const ALCchar* deviceName)
```

Comment:
I think I'd rather all actual reconnecting work only happen on the main thread, and not in the pool thread or the device watcher thread. The device watcher thread can continue to check for changes, which the main thread can query (in something called by `love.event.pump`, maybe).

Having device changes happen on multiple / different threads increases the chances of subtle race condition bugs a lot, and makes them harder to track down, compared to keeping it all in one place on the main thread.